File: C++/VigenereCipher.cpp
#include <iostream>
#include <string>
#include <cctype>
#include <unistd.h>
using namespace std;

string inmsg, key, enkey, cipher;
string alpha = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
int msglen, keylen, mIndex[100], kIndex[100], cIndex[100];

void spin()
{
      char spinner[] = {'/', '-', '\\', '|'};
    for(int i = 0; i < sizeof(spinner); i++)
    {
        cout << spinner[i];
        fflush(stdout);
        sleep(1);
        cout << "\b";
    }
}

void spinner()
{
    int pos;
    int spaces;
    int x;
    string text = "\tVigenere Cipher\n";
    
    for(pos = 0; pos <= text.length(); pos++)
    {
        for(x = 0; x < pos; x++)
        {
                cout << text[x];
        }
        if(pos+1 <= text.length())
        {
            spin();
            cout << "\r";
        }
        
    }
}

void getmsg()
{
    int i, x, j;
    cout<<"Enter the message: ";
    getline(cin, inmsg);
    msglen=inmsg.length();
    for(j=0;j<msglen;j++)
    {
        inmsg[j] = toupper(inmsg[j]);
    }
    for(i=0;i<msglen;i++)
    {
        x=0;
        while(inmsg[i] != alpha[x])
        {
            x++;
        }
        mIndex[i] = x;
    }
}

void getkey()
{
    int i, j, x;
    cout<<"Enter the key: ";
    cin>>key;
    cin.ignore();
    keylen=key.length();
    enkey=inmsg;
    for(i=0; i<msglen; i++)
    {
        enkey[i]=key[i%keylen];
    }
    
    for(int k=0;k<msglen;k++)
    {
        enkey[k] = toupper(enkey[k]);
    }
    for(j=0;j<msglen;j++)
    {
        x=0;
        while(enkey[j] != alpha[x])
        {
            x++;
        }
        kIndex[j] = x;
    }
}
void encrypt()
{
    getmsg();
    getkey();
    cipher=inmsg;
    int x;
    /*** Do Cipher ***/
    cout<<"\nCipher: \n";
    for(int i=0;i<msglen;i++)
    {
        cIndex[i]=(mIndex[i]+kIndex[i])%26;
        x = cIndex[i];
        cipher[i] = alpha[x];
    }
    cout<<cipher;
}

void decrypt()
{
    getmsg();
    getkey();
    cipher=inmsg;
    int x;
    /*** Do Decipher ***/
    cout<<"\nDeciphered: \n";
    for(int i=0;i<msglen;i++)
    {
        cIndex[i]=(mIndex[i]-kIndex[i])%26;
        x = cIndex[i];
        cipher[i] = alpha[x];
    }
    cout<<cipher;
}

int main()
{
    int choice;
    spinner();
    cout<<"1. Encrypt"<<endl;
    cout<<"2. Decrypt"<<endl;
    cout<<"Enter choice: ";
    cin>>choice;
    cin.ignore();
    switch(choice)
    {
        case 1:
        encrypt();
        break;
        
        case 2:
        decrypt();
        break;
    }
}

File: Python/VigenereCipher.py
import sys, time, threading

done = False
alpha = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
mIndex = ""
kIndex = ""
inMsg = ""
enKey = ""

def spin_cursor():
    while True:
        for cursor in '|/-\\':
            sys.stdout.write(cursor)
            sys.stdout.flush()
            time.sleep(0.1) # adjust this to change the speed
            sys.stdout.write('\b')
            global done
            if done:
                return

# start the spinner in a separate thread
def spinner():
	global done
	spin_thread = threading.Thread(target=spin_cursor)
	spin_thread.start()

	text = "    **** Vigenere Cipher ****      "
	textArray = list(text)
	# do some more work in the main thread, or just sleep:
	# time.sleep(1)

	for i in range(len(textArray)):
		sys.stdout.write('\r')
		for x in range(i):
			sys.stdout.write(textArray[x])
		time.sleep(0.2)
		
	time.sleep(1)
	# tell the spinner to stop, and wait for it to do so;
	# this will clear the last cursor before the program moves on
	done = True
	spin_thread.join()
	sys.stdout.write("\r")
	sys.stdout.write(text)
	# continue with other tasks
	sys.stdout.write("\n\n")

def getMsg():
	global alpha
	global mIndex
	
	mIndex = ""
	inMsg = ""
	
	inMsg = input("\n\nEnter the message: ").upper()
	
	for i in inMsg:
		for x in alpha:
			if i == x:
				mIndex += x

def getKey():
	global mIndex
	global kIndex
	global alpha
	
	kIndex = ""
	enKey = ""
	
	inKey = input("Enter the key: ").upper()
	for i in range(len(mIndex)):
		enKey += inKey[i%len(inKey)]
	
	for j in enKey:
		for k in alpha:
			if j == k:
				kIndex += k

def encrypt():
	getMsg()
	getKey()
	
	global kIndex
	global mIndex
	global alpha
	cipher = ""
	
	for i in range(len(mIndex)):
		x = (ord(mIndex[i]) + ord(kIndex[i])) % 26
		cipher += alpha[x]
	print("The encoded message: " + cipher + "\n\n")
	menu()

def decrypt():
	getMsg()
	getKey()
	
	global kIndex
	global mIndex
	global alpha
	cipher = ""
	
	for i in range(len(mIndex)):
		x = (ord(mIndex[i]) - ord(kIndex[i])) % 26
		cipher += alpha[x]
	print("The decoded message: " + cipher + "\n\n")
	menu()

def notOption():
	print("\nNot a choice\n")
	time.sleep(2)
	menu()

def exitProgram():
	raise SystemExit()

	
# Main Menu
def menu():
	print("1. Encrypt")
	print("2. Decrypt")
	print("3. Exit")
	choice = input("Enter choice: ")
	
	select = {'1': encrypt, '2': decrypt, '3': exitProgram}
	if choice in select:
		select[choice]()
	else:
		notOption()

### Main Area ###
spinner()
menu()